[RFC-3086] Add a new `concat` metavar expr

c410-f3r opened the pull request:
While investigating the blockers for the stabilization of #83527, I figured it would probably be nice to have a way to concatenate identifiers as well as literals without having to use third-parties or `concat_idents!`.

```rust
macro_rules! create_things {
    ( $lhs:ident ) => {
        struct ${concat(lhs, _separated_idents_in_a_struct)} {
            foo: i32,
            ${concat(lhs, _separated_idents_in_a_field)}: i32,
        }

        mod ${concat(lhs, _separated_idents_in_a_module)} {
            pub const FOO: () = ();
        }

        fn ${concat(lhs, _separated_idents_in_a_fn)}() {}
    };
}

create_things!(look_ma);

fn main() {
    look_ma_separated_idents_in_a_fn();

    let _ = look_ma_separated_idents_in_a_module::FOO;

    let _ = look_ma_separated_idents_in_a_struct {
        foo: 1,
        look_ma_separated_idents_in_a_field: 2,
    };
}
```

The intention here is to allow further experimentation of a nightly feature to gather more feedback. If such thing is not desired, needs a RFC or anything else, feel free to close this PR.

Implementation-wise, there are some things that still need to be tweaked.

* https://internals.rust-lang.org/t/macro-meta-functions/16743?u=cad97
* https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/macros.20in.20ident.20position
* https://github.com/rust-lang/rust/issues/29599
* https://github.com/rust-lang/rust/issues/13294
* https://github.com/rust-lang/rust/issues/50408

c410-f3r committed:
[RFC-3086] Add a new `concat` metavar expr

c410-f3r commented:
cc @petrochenkov in case you want to add something

c410-f3r commented:
Well, no-one seemed interested so I will close this PR. At least I tried 🤷

fmease commented:
> Well, no-one seemed interested

I think there's a chance that you can bring this new meta-var expr to the attention of more people by opening a new Zulip topic in the T-lang stream for example.

c410-f3r commented:
Hum... Some of my previous attempts trying to introduce new features resulted in a simple "no" or a feedback delay of +1 year, which doesn't give me the motivation to continue pursuing this PR.

In order to get more changes of success, the overall plan is to write a RFC addressing the remaining concerns of #83527 as well as including any other  useful meta variable expression (including `concat`).

oli-obk commented:
I don't know who's in charge of this feature, but I don't know enough about it to review this. I didn't even know there was an RFC and that it was accepted. I should still not have left the PR in limbo, I apologize for that.

Looking through the tracking issue, there seems to be no one really in charge of getting the feature implemented. While I can review the implementation, I do not want to make decisions on the language side of a feature I don't really know and don't have the time to get to know.

c410-f3r commented:
> I don't know who's in charge of this feature, but I don't know enough about it to review this. I didn't even know there was an RFC and that it was accepted. I should still not have left the PR in limbo, I apologize for that.

Oh, no need to apologize! My concerns were directed towards the big picture and such HR situation is common among open source projects. More like a feature than a bug :)

> Considering that (from the tracking issue discussion), you are the expert on this feature, you can just make feature decisions and we'll get it reviewed from the impl side. Worst case this will make someone from the compiler team disagree with the changes enough to step up laughing

It is much easier and faster to write a PR describing the intention behind the code than elaborating a formal RFC but I am still not sure because different persons showed different opinions in the tracking issue. Let me collect more material to figure things out.

fmease commented:
In any case, I just want to say thanks for working on this! :)
